### PR TITLE
ABI compat check: detect unadjusted ABI mismatches

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

File: compiler/rustc_target/src/abi/call/mod.rs [hunk 1]
```diff
@@ -745,10 +745,25 @@ impl<'a, Ty> ArgAbi<'a, Ty> {
...
+    {
         // Ideally we'd just compare the `mode`, but that is not enough -- for some modes LLVM will look
         // at the type.
-        self.layout.eq_abi(&other.layout) && self.mode.eq_abi(&other.mode)
+        self.layout.eq_abi(&other.layout) && self.mode.eq_abi(&other.mode) && {
+            // `fn_arg_sanity_check` accepts `PassMode::Direct` for some aggregates.
+            // That elevates any type difference to an ABI difference since we just use the
+            // full Rust type as the LLVM argument/return type.
+            if matches!(self.mode, PassMode::Direct(..))
+                && matches!(self.layout.abi, Abi::Aggregate { .. })
```
Comment (petrochenkov) [inline compiler/rustc_target/src/abi/call/mod.rs:759]: Are these `PassMode::Direct` concerns irrelevant for non-aggregates?

**Comment by petrochenkov** [comment]
This looks reasonable to me, but I may be missing some of the necessary context.
r=me, unless you want someone else to review.
@rustbot author

File: compiler/rustc_target/src/abi/call/mod.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_target/src/abi/call/mod.rs:759]: Yes -- for non-aggregates, codegen uses the `layout.abi` information to compute how the argument should be passed, which has already been checked to be sufficiently compatible. It is only with `Aggregate` that the underlying Rust type "leaks through".

**Pull request merged**